core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -403,42 +406,79 @@ private void logFailedEvents(final List<BatchItem> batch) {
+        final List<NakadiRecordMetadata> errors = new LinkedList<>();
+        int publishingEventIdx = 0;
         try {
-            final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
-                    nakadiRecord.getTopic(),
-                    nakadiRecord.getPartition(),
-                    nakadiRecord.getEventKey(),
-                    nakadiRecord.getData());
-
-            // fixme avro do not forget EventOwnerHeader, set via metadata?
-            // for the internal KPI evnets it is fine not to have it
-
-            producerRecord.headers().add(
-                    NakadiRecord.HEADER_FORMAT,
-                    nakadiRecord.getFormat());
-
-            producer.send(producerRecord, ((metadata, exception) -> {
-                if (null != exception) {
-                    LOG.warn("Failed to publish to kafka topic '{}' event-type '{}'",
-                            nakadiRecord.getTopic(), nakadiRecord.getEventType(), exception);
-                    if (isExceptionShouldLeadToReset(exception)) {
-                        kafkaFactory.terminateProducer(producer);
+            for (final NakadiRecord nakadiRecord : nakadiRecords) {
+                publishingEventIdx++;
+                final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
+                        topic,
+                        nakadiRecord.getPartition(),
+                        nakadiRecord.getEventKey(),
+                        nakadiRecord.getData());
+
+                // fixme avro do not forget EventOwnerHeader, set via metadata?
+                // for the internal KPI evnets it is fine not to have it
+
+                producerRecord.headers().add(
+                        NakadiRecord.HEADER_FORMAT,
+                        nakadiRecord.getFormat());
+                producer.send(producerRecord, ((metadata, exception) -> {
+                    latch.countDown();
+                    if (null != exception) {
+                        errors.add(new NakadiRecordMetadata(
+                                nakadiRecord.getEventType(),
+                                nakadiRecord.getPartition(),
+                                nakadiRecord.getData(),
+                                exception));
+                        if (isExceptionShouldLeadToReset(exception)) {
+                            kafkaFactory.terminateProducer(producer);
+                        }
                     }
-                }
-            }));
-        } catch (final InterruptException e) {
+                }));
+            }
+
+            latch.await(createSendTimeout(), TimeUnit.MILLISECONDS);
+        } catch (final InterruptException | InterruptedException e) {
             Thread.currentThread().interrupt();
-            LOG.error("Error publishing message to kafka", e);
+            markUnpublishedWithError(
+                    nakadiRecords.listIterator(publishingEventIdx),
```

Comment:
Does this assume that the publishing status of all the events before the `publishingEventIdx` is known?  Such use of the `LinkedList<>` also requires external synchronization, as per the docs note: 

`If multiple threads access a linked list concurrently, and at least one of the threads modifies the list structurally, it must be synchronized externally.`
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/LinkedList.html

Reply:
yep, good point, will address it